### ISSUE TITLE
Can't locate Foo/Bar.pm isn't a helpful message

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8637]
Migrated from [rt.perl.org#40516](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=40516) (status was 'resolved')

Searchable as RT40516$

**Comment by p5pRT** [comment]
### From @perlpunk

#### Created by @perlpunk



If you do 'use Foo&#8203;::Bar' and get the message 'Can't locate Foo/Bar\.pm
in @&#8203;INC' this message doesn't tell newbies what to do\. They don't see
the connection from Foo&#8203;::Bar to Foo/Bar\.pm \(because they don't know how
modules are internally stored\)\.
There are many questions in perl forums that ask about this issue\.
Usually you can solve this by answering that that means Foo&#8203;::Bar
isn't installed and link to a site which explains how to install modules\.

Could the message 'Can't locate Foo/Bar\.pm'be extended to something like
'\.\.\. \(This means that Foo&#8203;::Bar probably isn't installed on this machine\)'?
I was told in the perl5porters channel that at the point of the error message
the information of the module name might already be lost\, but maybe there
is a way to get it\.

\(Note&#8203;: Also 'use diagnostics' doesn't help much here\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl v5.8.5 - Fri Oct  1 23:29:33 UTC 2004
It is being executed now by  Perl v5.8.5 - Fri Oct  1 23:24:00 UTC 2004.

Site configuration information for perl v5.8.5:

Configured by abuild at Fri Oct  1 23:24:00 UTC 2004.

Summary of my perl5 (revision 5 version 8 subversion 5) configuration:
  Platform:
    osname=linux, osvers=2.6.8.1, archname=i586-linux-thread-multi
    uname='linux g168 2.6.8.1 #1 smp thu jul 1 15:23:45 utc 2004 i686 i686 i386 gnulinux '
    config_args='-ds -e -Dprefix=/usr -Dvendorprefix=/usr -Dinstallusrbinperl -Dusethreads -Di_db -Di_dbm -Di_ndbm -Di_gdbm -Duseshrplib=true -Doptimize=-O2 -march=i586 -mcpu=i686 -fmessage-length=0 -Wall -Wall -pipe'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -march=i586 -mcpu=i686 -fmessage-length=0 -Wall -Wall -pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe'
    ccversion='', gccversion='3.3.4 (pre 3.3.5 20040809)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =''
    libpth=/lib /usr/lib /usr/local/lib
    libs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.3'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/lib/perl5/5.8.5/i586-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared'

Locally applied patches:
    


@INC for perl v5.8.5:
    /usr/lib/perl5/5.8.5/i586-linux-thread-multi
    /usr/lib/perl5/5.8.5
    /usr/lib/perl5/site_perl/5.8.5/i586-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.5
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.5/i586-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.5
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.5:
    HOME=/home/tina
    LANG=de_DE@euro
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/tina/bin:/usr/local/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/games:/opt/gnome/bin:/opt/kde3/bin:/usr/lib/jvm/jre/bin
    PERLDOC=-i
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @Juerd

Tina skribis 2006\-10\-12  3&#8203;:06 \(\-0700\)&#8203;:
> If you do 'use Foo&#8203;::Bar' and get the message 'Can't locate Foo/Bar\.pm
> in @&#8203;INC' this message doesn't tell newbies what to do\. They don't see
> the connection from Foo&#8203;::Bar to Foo/Bar\.pm \(because they don't know how
> modules are internally stored\)\.

Many newbies do get part of the message\, but they think it says&#8203;: copy a
Foo/Bar\.pm into some directory of @&#8203;INC\. And that's how many people
"install" modules\, causing quite a lot of traffic on the diverse help
platforms when a module happens to use something other than pure Perl\.

> Could the message 'Can't locate Foo/Bar\.pm'be extended to something like
> '\.\.\. \(This means that Foo&#8203;::Bar probably isn't installed on this machine\)'?
> I was told in the perl5porters channel that at the point of the error message
> the information of the module name might already be lost\, but maybe there
> is a way to get it\.

Some heuristics would suffice\, I think\. For a path that has a slash in
it\, translate to :&#8203;: notation and suggest the module\. For all other
paths\, just keep including the \.pm\.

> \(Note&#8203;: Also 'use diagnostics' doesn't help much here\)

I'm afraid that most people don't know about "use diagnostics" and
"perldiag"\. The error message itself needs to be clear\.

I actually think the module complaint should come \*before\* the path&#8203;:

&nbsp;    "Can't load module Foo&#8203;::Bar \(no Foo/Bar\.pm in @&#8203;INC\)"

"Can't locate loadable object for module Foo&#8203;::Bar in @&#8203;INC" curiously
does have the :&#8203;: notation in it\, and also has the user friendly text
first\.
\-\- 
korajn salutojn\,

&nbsp;  juerd waalboer&#8203;:  perl hacker  \<juerd@&#8203;juerd\.nl>  \<http&#8203;://juerd\.nl/sig>
&nbsp;  convolution&#8203;:     ict solutions and consultancy \<sales@&#8203;convolution\.nl>

Ik vertrouw stemcomputers niet\.
Zie \<http&#8203;://www\.wijvertrouwenstemcomputersniet\.nl/>\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 10/12/06\, Juerd \<juerd@&#8203;convolution\.nl> wrote&#8203;:
> Tina skribis 2006\-10\-12  3&#8203;:06 \(\-0700\)&#8203;:
> > If you do 'use Foo&#8203;::Bar' and get the message 'Can't locate Foo/Bar\.pm
> > in @&#8203;INC' this message doesn't tell newbies what to do\. They don't see
> > the connection from Foo&#8203;::Bar to Foo/Bar\.pm \(because they don't know how
> > modules are internally stored\)\.
>
> Many newbies do get part of the message\, but they think it says&#8203;: copy a
> Foo/Bar\.pm into some directory of @&#8203;INC\. And that's how many people
> "install" modules\, causing quite a lot of traffic on the diverse help
> platforms when a module happens to use something other than pure Perl\.
>
> > Could the message 'Can't locate Foo/Bar\.pm'be extended to something like
> > '\.\.\. \(This means that Foo&#8203;::Bar probably isn't installed on this machine\)'?
> > I was told in the perl5porters channel that at the point of the error message
> > the information of the module name might already be lost\, but maybe there
> > is a way to get it\.
>
> Some heuristics would suffice\, I think\. For a path that has a slash in
> it\, translate to :&#8203;: notation and suggest the module\. For all other
> paths\, just keep including the \.pm\.
>
> > \(Note&#8203;: Also 'use diagnostics' doesn't help much here\)
>
> I'm afraid that most people don't know about "use diagnostics" and
> "perldiag"\. The error message itself needs to be clear\.
>
> I actually think the module complaint should come \*before\* the path&#8203;:
>
>     "Can't load module Foo&#8203;::Bar \(no Foo/Bar\.pm in @&#8203;INC\)"
>
> "Can't locate loadable object for module Foo&#8203;::Bar in @&#8203;INC" curiously
> does have the :&#8203;: notation in it\, and also has the user friendly text
> first\.

Maybe use should localize a global temporarily to make the module name
available to require\. So instead of use Foo&#8203;::Bar; being

BEGIN\{
&nbsp;   require \.\.\.\.
&nbsp;   import \.\.\.
\}

it becomes

BEGIN\{
&nbsp;   local $^MODULE="Foo&#8203;::Bar";
&nbsp;   require \.\.\.\.
&nbsp;   import \.\.\.\.
\}

Then extensions could use it themselves to override the message as
they felt appropriate before calling require\.

This would also work around the problem that the original name is lost
by the time PP\(require\) is executed\, which is the routine responsible
for this message\.

I had a quick look into doing this but its not a part of the perl
internals im very familiar with\.

Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @jrockway

Juerd wrote&#8203;:
> Tina skribis 2006\-10\-12  3&#8203;:06 \(\-0700\)&#8203;:
>> If you do 'use Foo&#8203;::Bar' and get the message 'Can't locate
>> Foo/Bar\.pm in @&#8203;INC' this message doesn't tell newbies what to do\.
>> They don't see the connection from Foo&#8203;::Bar to Foo/Bar\.pm \(because
>> they don't know how modules are internally stored\)\.
> 
> Many newbies do get part of the message\, but they think it says&#8203;: copy
> a Foo/Bar\.pm into some directory of @&#8203;INC\. And that's how many people 
> "install" modules\, causing quite a lot of traffic on the diverse help
>  platforms when a module happens to use something other than pure
> Perl\.

I disagree that "use diagnostics" doesn't help\.  I think the message is
pretty clear&#8203;:

\(F\) You said to do \(or require\, or use\) a file that couldn't be
found\. Perl looks for the file in all the locations mentioned in @&#8203;INC\,
unless the file name included the full path to the file\.  Perhaps you
need to set the PERL5LIB or PERL5OPT environment variable to say where
the extra library is\, or maybe the script needs to add the library name
to @&#8203;INC\.  Or maybe you just misspelled the name of the file\.  See
perlfunc/require and lib\.

Perhaps mentioning CPAN \(and cpan\) would help?  "You might also try
installing the module from CPAN by typing 'cpan Foo&#8203;::Bar'\."  This opens
up a can of worms\, of course\.  The user might not have permission to
install modules\.  The user might be better served by using their
distribution's installation method\.  The user might not be able to
invoke cpan from the command line \(Activestate?\)\.  The user might not
have make\.  The user might not have a C compiler\.  And so on :\)

However\, referring the user to some relevant documentation might be a
good idea\.  perlfunc and lib aren't helpful in this case \-\- cpan and
perlmodinstall would make more sense\.  My attitude is that if you want
to program\, you need to know how to read and think and experiment\.
Pointing the user to documentation\, therefore\, is an excellent idea\.
Magically solving the problem for them is just going to lead to worse
problems down the line\, like $foo = param\('foo'\); print \`$foo bar\`; :\)

Regards\,
Jonathan Rockway

\-\- 
package JAPH;use Catalyst qw/\-Debug/;\($;=JAPH\)\->config\(name => do \{
$\,\.=reverse qw\[Jonathan tsu rehton lre rekca Rockway\]\[$\_\]\.\[split //\,
";$;"\]\->\[$\_\]\.q; ;for 1\.\.4;$\,=~s;^\.;;;$\,\}\);$;\->setup;

**Comment by p5pRT** [comment]
### From @jimc

Jonathan Rockway wrote&#8203;:
> Juerd wrote&#8203;:
>   
>> Tina skribis 2006\-10\-12  3&#8203;:06 \(\-0700\)&#8203;:
>>     
>>> If you do 'use Foo&#8203;::Bar' and get the message 'Can't locate
>>> Foo/Bar\.pm in @&#8203;INC' this message doesn't tell newbies what to do\.
>>> They don't see the connection from Foo&#8203;::Bar to Foo/Bar\.pm \(because
>>> they don't know how modules are internally stored\)\.
>>>       
>> Many newbies do get part of the message\, but they think it says&#8203;: copy
>> a Foo/Bar\.pm into some directory of @&#8203;INC\. And that's how many people 
>> "install" modules\, causing quite a lot of traffic on the diverse help
>>  platforms when a module happens to use something other than pure
>> Perl\.
>>     
>
> I disagree that "use diagnostics" doesn't help\.  I think the message is
> pretty clear&#8203;:
>
>   
it doesnt help when they dont use it\, which is common\.

I wouldnt mind seeing the 1st 3 errors verbosified \(with the rest left 
as is\, to not overwhelm\)
or maybe \-W should spew diagnostics too\.

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'stalled'

**Comment by p5pRT** [comment]
### From @cpansprout

This is now resolved by commit f7ee53b552\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'stalled' to 'resolved'
